crypto/signify.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// signFile reads the contents of an input file and signs it (in armored format)
+// with the key provided, placing the signature into the output file.
+
+package crypto
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"crypto/ed25519"
+)
+
+var (
+	errInvalidKeyHeader = errors.New("Incorrect key header")
+	errInvalidKeyLength = errors.New("invalid, key length != 104")
+)
+
+func readSKey(key []byte) (ed25519.PrivateKey, error) {
+	if len(key) != 104 {
+		return nil, errInvalidKeyLength
+	}
+
+	if string(key[:2]) != "Ed" {
+		return nil, errInvalidKeyHeader
+	}
+
+	return ed25519.PrivateKey(key[40:]), nil
+
+}
+
+func isCommentOnlyOneLine(comment string) bool {
+	firstCRIndex := strings.IndexByte(comment, 13)
+	firstLFIndex := strings.IndexByte(comment, 10)
+	return (firstCRIndex >= 0 && firstCRIndex < len(comment)-1) || (firstLFIndex >= 0 && firstLFIndex < len(comment)-1)
```

Comment:
Afaict, this allows arbitrarily many `CR` or `LF` characters to be placed anywhere? 
A one-liner to me means that it contains exactly one or zero `CR` _or_ `CRLF`, and that is placed at the end.

Reply:
the code is correct, it's the function name that is wrong. It should be `commentMoreThanOneLine`. If this returns `true`, then an error will be returned.

Reply:
The go-minisign verifier does this: 
```
	lines := strings.SplitN(in, "\n", 4)
```
So it totally ignores `CR` and counts only `LF`.

Reply:
Maybe it's not true in Go, but `\n` is OS-dependent in C, so it could be `CR`, `LF`, or both (don't have Windows/MacOS, can't check that). Maybe @fjl can chime in here? I could remove the check for `LF` but I want to make sure that it's going to work on Windows/MacOS.

Reply:
On unix-systems, `\n` (`LF`) is used, whereas windows uses `\r\n` (`CRLF`). We should we stick to whatever signify and_or minisign does, and hopefully both of them behaves the same way. 

If they allow "lalalal \r lolo \r foo bar\n" then we should do that too

Reply:
> On unix-systems, `\n` (`LF`) is used, whereas windows uses `\r\n` (`CRLF`). We should we stick to whatever signify and_or minisign does, and hopefully both of them behaves the same way.
> 
> If they allow "lalalal \r lolo \r foo bar\n" then we should do that too

this is what I said, and MacOS only uses `CR` in some modes. However, I guess `go-minisign` will behave the same on every platform, which means that it will expect `CR` only. I just tested `signify-openbsd` and it's ignoring `LF`. So I assume there is no need to test on Windows/MacOS after all.

Reply:
oh right, yes, a C program in windows will require `\r\n`, of course. My bad. Monday mornings. :sleeping: Well, I just pushed the code to do that anyway.

Reply:
>   just tested signify-openbsd and it's ignoring LF. So I assume there is no need to test on Windows/MacOS after all.

Agreed

> this is what I said

No, you said that `\n` is OS-depdendent, which is not true. It's very well defined :-p

Reply:
> No, you said that `\n` is OS-depdendent, which is not true. It's very well defined :-p

Someone doesn't refresh before posting :D

Reply:
> Someone doesn't refresh before posting :D

Yup, I hate that -- sometimes GH just plops in whatever new content there is, and sometimes you have to manually reload pages. Solly